### PR TITLE
Fix rustdoc run failures by shutting down definitely some lints

### LINKED ISSUE #51661: Cargo doc regression in crates that include #![deny(unused_imports)]
```rust
#![deny(unused_imports)]

use std::fs::File;

pub fn f() {
    let _: File;
}
```

`cargo doc` of the above lib.rs works up through `cargo +nightly-2018-06-17 doc`. As of `cargo +nightly-2018-06-18 doc` it fails with:

```console
$ cargo +nightly-2018-06-18 doc
 Documenting unused-imports v0.0.0
error: unused import: `std::fs::File`
 --> src/lib.rs:3:5
  |
3 | use std::fs::File;
  |     ^^^^^^^^^^^^^
  |
note: lint level defined here
 --> src/lib.rs:1:9
  |
1 | #![deny(unused_imports)]
  |         ^^^^^^^^^^^^^^

error: Compilation failed, aborting rustdoc

error: Could not document `unused-imports`.
```

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #51661.

cc @oli-obk @arielb1 @eddyb

### CODE REVIEW

File: src/librustdoc/core.rs [hunk 1]
```diff
@@ -235,6 +235,14 @@ pub fn run_core(search_paths: SearchPaths,
         let mut sess = session::build_session_(
             sessopts, cpath, diagnostic_handler, codemap,
         );
+
+        let shutdown_lints = [lint::builtin::UNUSED_IMPORTS,
```
Comment (oli-obk) [inline src/librustdoc/core.rs]: wait that's the wrong way around. I thought you were going to turn on the rustdoc lints even if lints are capped

Comment (GuillaumeGomez) [reply src/librustdoc/core.rs]: I do, but I need to deactivate some lints even if they're enabled.

Comment (oli-obk) [reply src/librustdoc/core.rs]: You can just turn the cap lints back on as it was before

**Comment by oli-obk** [comment]
We should discuss this in the compiler meeting, because adding/changing lints in the compiler can now randomly break rustdoc for ppl with any kind of `deny` on, even if that `deny` does not cause any fallout during normal compilation

File: src/librustdoc/core.rs [hunk 2]
```diff
@@ -235,6 +236,24 @@ pub fn run_core(search_paths: SearchPaths,
...
         );
+
+        lint::builtin::HardwiredLints.get_lints()
+                                     .into_iter()
+                                     .chain(rustc_lint::SoftLints.get_lints().into_iter())
+                                     .filter_map(|lint| {
+                                         if lint.name == warnings_lint_name ||
+                                            lint.name == intra_link_resolution_failure_name ||
+                                            lint.name == missing_docs {
+                                             None
```
Comment (oli-obk) [inline src/librustdoc/core.rs]: Took me a while to figure this out. So you are creating a list of *all* lints except for those you want to keep, and then you set all of them to `allow`.

Can you add a comment about this here?

Explanation about why I was confused:

I initially assumed we'd build a list of the lints we want to keep (so essentially `vec![warnings_lint_name, intra_link_resolution_failure_name, missing_docs]`, set `--cap-lints=allow` and then make sure that cap-lints doesn't override our specific set of lints.

r=me with comment

**Comment by pietroalbini** [comment]
Nominating for beta backport since this fixes a regression

**Comment by pietroalbini** [comment]
Marked as `beta-accepted` (#52313).

**Pull request merged**